song-server/src/main/java/org/icgc/dcc/song/server/service/UploadService.java

```diff
@@ -62,16 +66,52 @@ public Upload read(@NonNull String uploadId) {
     return uploadRepository.get(uploadId);
   }
 
-  private void create(@NonNull String studyId, @NonNull String uploadId, @NonNull String jsonPayload) {
-    uploadRepository.create(uploadId, studyId, Upload.CREATED, jsonPayload);
+  private void create(@NonNull String studyId, String analysisSubmitterId, @NonNull String uploadId,
+                      @NonNull String jsonPayload) {
+    uploadRepository.create(uploadId, studyId, analysisSubmitterId, Upload.CREATED, jsonPayload);
+  }
+
+  private void update(@NonNull String uploadId, @NonNull String jsonPayload) {
+    uploadRepository.update_payload(uploadId, Upload.UPDATED, jsonPayload);
   }
 
   @SneakyThrows
   public ResponseEntity<String> upload(@NonNull String studyId, @NonNull String payload, boolean isAsyncValidation) {
-    val uploadId = id.generate(IdPrefix.Upload);
     String analysisType;
+    String uploadId;
+    val status = JsonUtils.ObjectNode();
+    status.put("status","ok");
+
     try {
-      create(studyId, uploadId, payload);
+      val analysisSubmitterId=JsonUtils.readTree(payload).at("/analysisSubmitterId").asText();
+      List<String> ids;
+
+
+      if (analysisSubmitterId.equals("")) {
+        // Our business rules say that we always want to create a new record if no analysisSubmitterId is set,
+        // even if the rest of the content is duplicated.
+        ids = Collections.emptyList();
+      } else {
+        ids = uploadRepository.findByBusinessKey(studyId, analysisSubmitterId);
+      }
+
+      if (ids.isEmpty()) {
```

Comment:
Is it possible that the `else` conditional branch from the previous `if` returns `null` for the ids? if so this would throw a `NPE`.